Sources/Basic/Path.swift

```diff
@@ -0,0 +1,473 @@
+/// tactic, and does not access the file system in any way.
+public struct RelativePath {
+    /// Private implementation details, shared with the AbsolutePath struct.
+    private let _impl: PathImpl
+    
+    /// Initializes the RelativePath from `str`, which must be a relative path
+    /// (which means that it must not begin with a path separator or a tilde).
+    /// An empty input path is allowed, but will be normalized to a single `.`
+    /// character.
+    ///
+    /// The path string will be normalized by:
+    /// - Collapsing `..` path components that aren't at the beginning
+    /// - Removing extraneous `.` path components
+    /// - Removing any trailing path separator
+    /// - Removing any redundant path separators
+    /// - Replacing a completely empty path with a `.`
+    ///
+    /// This string manipulation may change the meaning of a path if any of the
+    /// path components are symbolic links on disk.  However, the file system is
+    /// never accessed in any way when initializing an AbsolutePath.
+    public init(_ string: String) {
+        precondition(string.characters.first != PathSeparatorCharacter
+                  && string.characters.first != HomeDirectoryCharacter)
+        // Get a hold of the character view.
+        // FIXME: We need to investigate what is most performant here.
+        //        For example, is UTF-8 view more efficient, or unichars, etc?
+        let chars = string.characters
+        
+        // Split the character array into parts, folding components as we go.
+        // As we do so, we count the number of characters we'll end up with in
+        // the normalized string representation.
+        var parts: [String.CharacterView] = []
+        var capacity = 0
+        for part in chars.split(separator: PathSeparatorCharacter) {
+            switch part.count {
+              case 0:
+                // Ignore empty path components.
+                continue
+              case 1 where part.first == ".":
+                // Ignore `.` path components.
+                continue
+              case 2 where part.first == "." && part.last == ".":
+                // If at beginning, fall through to treat the `..` literally.
+                guard let prev = parts.last else {
+                    fallthrough
+                }
+                // If previous component is anything other than `..`, drop it.
+                if !(prev.count == 2 && prev.first == "." && prev.last == ".") {
+                    parts.removeLast()
+                    capacity -= prev.count
+                    continue
+                }
+                // Otherwise, fall through to treat the `..` literally.
+                fallthrough
+              default:
+                // Any other component gets appended.
+                parts.append(part)
+                capacity += part.count
+            }
+        }
+        capacity += max(parts.count - 1, 0)
```

Comment:
This isn't really code duplication except for a few lines.  I started off writing a single method that handled both absolute and relative paths, but because of differences in the semantics of `..` at the beginning of absolute vs relative paths, the unified code turned out to be a lot more complicated.  So I split it out.